src/Bicep.Core/TypeSystem/ArmFunctionReturnTypeEvaluator.cs

```diff
@@ -0,0 +1,207 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using Azure.Deployments.Expression.Expressions;
+using Bicep.Core.Diagnostics;
+using Bicep.Core.Extensions;
+using Newtonsoft.Json.Linq;
+
+namespace Bicep.Core.TypeSystem;
+
+public static class ArmFunctionReturnTypeEvaluator
+{
+    public static TypeSymbol Evaluate(
```

Comment:
Do you have a feel for how computationally expensive this is?

To put it politely - I'm not sure the Deployments expression evaluation logic is particularly optimized for efficiency 😄

Reply:
It's not free, but I think most of the added computation would be in converting from `TypeSymbol`s to `JToken`s to .NET types and back again. A function like `mul` is probably as simple as they get: it takes an array of `FunctionArgument`s, validates that there are exactly two of them, validates that they contain integers, converts them to longs, performs a checked invocation of the `*` operator on the longs, then converts the result back to a `JToken`. At a glance, the most expensive function is probably `replace`, and we're already performing type transformations on functions that operate on and return strings. The added expense would almost all be coming from evaluating operators, which I believe all boil down to fairly cheap functions.

Is there a good way to measure this?

Reply:
I should add that the thing we *aren't* doing now with string transformation functions is handling type unions. If `toLower` gets called on an argument with a type of `'Foo'|'Bar'`, the compiler will say the function returns a generic string rather than `'foo'|'bar'`. The third commit in this PR handles transformations of unions by generating all possible permutations of the supplied input arguments, which has a worst-case complexity of O(`n`!), where `n` is the already not so cheap cost of evaluating a single ARM expression.

Reply:
We could do some basic benchmarking here. Maybe grab all the checked in files, run n iterations of compiling the same file times m samples and see what the difference is in compile time (or type checking time) with and without this logic? (On one machine with similar load - taking m samples should even things out.)

Reply:
Alternative perf test would be to give this build to miqm and have him tell us how it runs with his workspace of 239084309483094 modules.

Reply:
Yeah, the handling of permutations was what prompted me to write the comment. Feels like that could potentially add up quickly, as processing sets of params with long lists of allowed values feels like it's going to come up somewhat frequently. E.g. thinking about an expression like `toLower('${region}-${environment}-${type}-${sku})` where each interpolation hole is an enum from an allowed-values-param

Reply:
The potential permutations are calculated upfront, so I could set an arbitrary limit to how many will be processed. Maybe if there are more than 10 possible permutations, the evaluator just falls back to the non-literal return type?

Reply:
For now, we can set the arbitrary limit of how many permutations to process to 1 (by dropping the union type support). No one has yet complained about functions like `toLower` not supporting union types.